Code/Tools/AssetProcessor/native/utilities/ApplicationManagerBase.cpp

```diff
@@ -431,79 +431,98 @@ void ApplicationManagerBase::DestroyPlatformConfiguration()
     }
 }
 
-void ApplicationManagerBase::InitFileMonitor()
+void ApplicationManagerBase::InitFileMonitor(AZStd::unique_ptr<FileWatcher> fileWatcher)
 {
+    m_fileWatcher = AZStd::move(fileWatcher);
+
     for (int folderIdx = 0; folderIdx < m_platformConfiguration->GetScanFolderCount(); ++folderIdx)
     {
         const AssetProcessor::ScanFolderInfo& info = m_platformConfiguration->GetScanFolderAt(folderIdx);
-        m_fileWatcher.AddFolderWatch(info.ScanPath(), info.RecurseSubFolders());
+        m_fileWatcher->AddFolderWatch(info.ScanPath(), info.RecurseSubFolders());
     }
 
     QDir cacheRoot;
     if (AssetUtilities::ComputeProjectCacheRoot(cacheRoot))
     {
-        m_fileWatcher.AddFolderWatch(cacheRoot.absolutePath(), true);
+        m_fileWatcher->AddFolderWatch(cacheRoot.absolutePath(), true);
     }
 
     if (m_platformConfiguration->GetScanFolderCount() || !cacheRoot.path().isEmpty())
     {
         const auto cachePath = QDir::toNativeSeparators(cacheRoot.absolutePath());
 
+        // For the handlers below, we need to make sure to use invokeMethod on any QObjects so Qt can queue
+        // the callback to run on the QObject's thread if needed.  The APM methods for example are not thread-safe.
+
         const auto OnFileAdded = [this, cachePath](QString path)
         {
             const bool isCacheRoot = path.startsWith(cachePath);
-            if (isCacheRoot)
+            if (!isCacheRoot)
             {
-                m_fileStateCache->AddFile(path);
-            }
-            else
-            {
-                m_assetProcessorManager->AssessAddedFile(path);
-                m_fileStateCache->AddFile(path);
-                AZ::Interface<AssetProcessor::ExcludedFolderCacheInterface>::Get()->FileAdded(path);
-                m_fileProcessor->AssetProcessor::FileProcessor::AssessAddedFile(path);
+                [[maybe_unused]] bool result = QMetaObject::invokeMethod(m_assetProcessorManager, "AssessAddedFile", Q_ARG(QString, path));
```

Comment:
I don't like using the string-based method invocation. You lose compile-time connection checks, and it is harder to find places where that method is used or called. It would be better to use a lambda here:
```suggestion
                [[maybe_unused]] bool result = QMetaObject::invokeMethod(m_assetProcessorManager, [this, path]
                {
                    m_assetProcessorManager->AssessAddedFile(path);
                });
```